### ISSUE TITLE
Automatically add type parameters for unparameterized fields

### ISSUE THREAD

**Issue by colinxs** [JuliaLang/julia#31231]
Consider the following two structs:
```
struct Foo
x
end

struct Baz{T}
x::T
end
```
As discussed in the Performance Tips sections, functions that operate on `Baz` will outperform those that operate on `Foo` as the type of x is declared in the definition of `Baz`, thus allowing functions to specialize on the parameterized type of `Baz`.

When prototyping new code in a research setting, I often find myself adding and removing fields. This requires me to add and remove type parameters for every change to the type definition if I want operations on the type to be performant, which can be cumbersome if there are lots of fields.

Ideally, I would like for those annotations to happen automatically for untyped fields along the lines of:
```
@autotype struct Foo
x
y
z
end
```
which would exactly correspond to:
```
struct Foo{T1,T2,T3}
x::T1
y::T2
z::T3
end
```

Does such a solution exist?

Thank you in advance (and apologies if this is the incorrect place to ask such a question)!

-Colin

**Comment by jpsamaroo** [comment]
I think this is probably out of scope for Base or a Stdlib, but this is certainly fair game for a package. It's also the best way to "incubate" such functionality before making consideration to include it in Julia itself.


To your question at the end, https://discourse.julialang.org/ would have also been an acceptable place to ask this question.

**Comment by JeffBezanson** [comment]
I have thought about doing this, since it would make structs "fast by default" the same way functions are. Using a macro for this is of course fine, but there are some issues with doing it in the language by default, which is why we haven't done it.

First, seeing `struct Foo` makes you think `Foo` will be the concrete type of all instances. But surprise, `Foo` is actually abstract, and uses like `Vector{Foo}` will be inefficient. Similarly, `Type{Foo}` in method signatures will not work as expected; you might need `Type{<:Foo}` instead.

Second, it's useful to the compiler to be able to *not* specialize on something for a change :) Adding these parameters could make the compiler do a lot of work the programmer didn't ask for.

Anyway, this is just to give some background on why we don't do this already, not to say this macro is a bad idea.

**Comment by colinxs** [comment]
@jpsamaroo I agree that it's probably out of scope for Base/stdlib. For similar questions in the future I'll post to discourse :)

@JeffBezanson It makes sense why this isn't the default. Ideally it would be nice to be able to switch easily between the two (fully typed and untyped) depending on whether or not you want to favor compile-time or runtime speed. 

I took a look at Parameters.jl and QuickTypes.jl, two packages which approach this idea. Parameters doesn't seem to add any typing information, while QuickTypes makes a complete departure from the normal struct definition syntax. I'll likely put something together to fill this gap as it would be useful for my own work and leave a link here for reference (in case anyone else has a similar question).

Feel free to close this issue unless you see any reason to keep it open!

Thanks again for all your input.
Colin

**Comment by mauro3** [comment]
In Parameters.jl, there is the `@deftype`:
```
@with_kw struct A @deftype Float64
  a # ::Float64 is added
end
# also works
@with_kw struct B{T} @deftype T
  b # ::T
end
```
This covers almost all of my cases and I find it ok, to be explicit the other times.  Also, consider using NamedTuples for protyping (fast by default) and then once settled move to `struct`s; that also helps with the Revise.jl work-flow.

**Comment by lassepe** [comment]
I still think it would be nice to have a macro for this that is more flexible than `@deftype` to support heterogeneous field types. The same way that we now have `Base.@kwdef`. In particular, I often find my self introducing custom structs since they are one of the few principled ways to document the data layout (and have custom dispatch behavior). This something that is hard to achieve with`NamedTuples` alone (although I do like then for prototyping to work with Revise). Or is there some other recommended workflow for this?

**Comment by lassepe** [comment]
Just to give an example of something where this just bit me:

```julia
struct TrajectoryGenerator
    parametrization
    T,
    n,
    num_equality,
    parametric_cost,
    parametric_cost_grad,
    parametric_cost_jac,
    parametric_cons,
    lb,
    ub,
    jac_primals,
    jac_params,
    hess_primals
end
```
Here the field types range from Integers, over callables (which may not necessarily be a function but a callable struct), to vectors and matrices. I would very much prefer not to come up with 13 random type parameter names in order to make this struct fast.

**Comment by vtjnash** [comment]
It could be, amusingly, an application of ComputedFieldTypes:
```julia
julia> using ComputedFieldTypes

julia> @computed struct TrajectoryGenerator
           parametrization::typeof(parametrization)
           T::typeof(T)
       end

julia> TrajectoryGenerator(1, 2.0)
TrajectoryGenerator{Int64, Float64}(1, 2.0)
```

**Comment by lassepe** [comment]
That is a great idea which I quite like a lot. My only remaining gripe then would be that `@computed` does not compose with `Base.@kwdef` but I guess that's a different issue.

**Comment by vtjnash** [comment]
I think I added that composition support very recently
